libp2p-floodsub/src/lib.rs

```diff
@@ -0,0 +1,640 @@
+
+    // List of topics we're subscribed to. Necessary in order to filter out messages that we
+    // erroneously receive.
+    subscribed_topics: RwLock<Vec<Topic>>,
+
+    // Sequence number for the messages we send.
+    seq_no: AtomicUsize,
+
+    // We keep track of the messages we received (in the format `(remote ID, seq_no)`) so that we
+    // don't dispatch the same message twice if we receive it twice on the network.
+    // TODO: the `HashSet` will keep growing indefinitely :-/
+    received: Mutex<FnvHashSet<(Vec<u8>, Vec<u8>)>>,
+}
+
+struct RemoteInfo {
+    // Sender to send data over the socket to that host.
+    sender: mpsc::UnboundedSender<BytesMut>,
+    // Topics the remote is registered to.
+    subscribed_topics: RwLock<FnvHashSet<TopicHash>>,
+}
+
+impl fmt::Debug for Inner {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        fmt.debug_struct("Inner")
+            .field("peer_id", &self.peer_id)
+            .field(
+                "num_remote_connections",
+                &self.remote_connections.read().len(),
+            )
+            .field(
+                "num_subscribed_topics",
+                &self.subscribed_topics.read().len(),
+            )
+            .field("seq_no", &self.seq_no)
+            .field("received", &self.received)
+            .finish()
+    }
+}
+
+impl<T, C> FloodSubController<T, C>
+where
+    T: MuxedTransport + 'static,                // TODO: 'static :-/
+    C: ConnectionUpgrade<T::RawConn> + 'static, // TODO: 'static :-/
+{
+    /// Builds a new controller for floodsub.
+    #[inline]
+    pub fn new(
+        upgrade: &FloodSubUpgrade,
+        swarm: SwarmController<T, C>,
+    ) -> FloodSubController<T, C> {
+        FloodSubController {
+            inner: upgrade.inner.clone(),
+            swarm: swarm,
+        }
+    }
+
+    /// Subscribe to a topic. When a node on the network sends a message for that topic, we will
+    /// likely receive it.
+    ///
+    /// It is not guaranteed that we receive every single message published on the network.
+    #[inline]
+    pub fn subscribe(&self, topic: &Topic) {
+        // This function exists for convenience.
+        self.subscribe_multi(iter::once(topic));
+    }
+
+    /// Same as `subscribe`, but subscribes to multiple topics at once.
+    #[inline]
+    pub fn subscribe_multi<'a, I>(&self, topics: I)
+    where
+        I: IntoIterator<Item = &'a Topic>,
+        I::IntoIter: Clone,
+    {
+        // This function exists for convenience.
+        self.sub_unsub_multi(topics.into_iter().map::<_, fn(_) -> _>(|t| (t, true)))
+    }
+
+    /// Unsubscribe from a topic. We will no longer receive any message for this topic.
+    ///
+    /// If a message was sent to us before we are able to notify that we don't want messages
+    /// anymore, then the message will be filtered out locally.
+    #[inline]
+    pub fn unsubscribe(&self, topic: &Topic) {
```

Comment:
Same as above, mention that `unsubscribe_multi` is preferred to multiple calls to `unsubscribe` and consider calling this `unsubscribe_many`.